src/main/scala/gemmini/CmdFSM.scala

```diff
@@ -0,0 +1,198 @@
+  // local state registers
+  //==========================================================================
+  val m              = RegInit(0.U(32.W))
+  val n              = RegInit(0.U(32.W))
+  val k              = RegInit(0.U(32.W))
+  val addr_a         = RegInit(0.U(xLen.W))
+  val addr_b         = RegInit(0.U(xLen.W))
+  val addr_c         = RegInit(0.U(xLen.W))
+  val addr_d         = RegInit(0.U(xLen.W))
+  val in_rshift      = RegInit(0.U(log2Up(accType.getWidth).W))
+  val acc_rshift     = RegInit(0.U(log2Up(accType.getWidth).W))
+  val relu6_lshift   = RegInit(0.U(log2Up(accType.getWidth).W))
+  val activation     = RegInit(0.U(2.W))
+  val repeating_bias = RegInit(0.U(1.W))
+
+  // Valid fields
+  val addr_ab_valid = RegInit(false.B)
+  val addr_cd_valid = RegInit(false.B)
+  val size0_valid = RegInit(false.B)
+  val size1_valid = RegInit(false.B)
+  val config_ex_valid = RegInit(false.B)
+  val bias_valid = RegInit(false.B)
+
+  // pass CSR status bits to tiler (TODO: fix this api)
+  val status = Reg(new MStatus)
+  status := DontCare
+
+  //==========================================================================
+  // Combinational Output Defaults 
+  //==========================================================================
+  io.cmd.ready         := false.B
+  io.tiler.valid       := false.B
+  io.tiler.bits.status := status
+  io.flush_retry       := false.B
+  io.flush_skip        := false.B
+
+  io.tiler.bits.m              := m
+  io.tiler.bits.n              := n
+  io.tiler.bits.k              := k
+  io.tiler.bits.addr_a         := addr_a
+  io.tiler.bits.addr_b         := addr_b
+  io.tiler.bits.addr_c         := addr_c
+  io.tiler.bits.addr_d         := addr_d
+  io.tiler.bits.in_rshift      := in_rshift
+  io.tiler.bits.acc_rshift     := acc_rshift
+  io.tiler.bits.relu6_lshift   := relu6_lshift
+  io.tiler.bits.activation     := activation
+  io.tiler.bits.repeating_bias := repeating_bias
+
+  // we block if we have accepted the COMPUTE_ALL command, but the tiler has
+  // not started executing it yet.
+  io.busy := (state === s_EX_PENDING)
+
+  //==========================================================================
+  // FSM 
+  //==========================================================================
+  def reset_and_listen(): Unit = {
+    // Reset all data-validity
+    addr_ab_valid := false.B
+    addr_cd_valid := false.B
+    size0_valid := false.B
+    size1_valid := false.B
+    config_ex_valid := false.B
+    bias_valid := false.B
+    // And go back to listening for commands
+    state := s_LISTENING
+  }
+
+  when (state === s_EX_PENDING) {
+    // Pending s_EX ongoing
+    // Wait for tiling/ execution to complete,
+    // let any further commands queue up
+    io.tiler.valid := true.B
+    when (io.tiler.fire()) {
+      state := s_LISTENING
+    }
+  }.elsewhen (state === s_ERROR) {
+    // In s_ERROR state - only update based on RESET commands
```

Comment:
When do errors occur exactly? Does that imply a software bug?

Reply:
right now i don't think this code path is even reachable since i added the `is_cisc_funct` switch in the Controller module. 

but this originally put gemmini in an error state if the software issued a risc command or any other invalid command to the cisc frontend. i thinks its probably fine to leave it in ... the hardware is not incorrect by its presence.